javascript/0103-binary-tree-zigzag-level-order-traversal.js

```diff
@@ -0,0 +1,49 @@
+/**
+ * Definition for a binary tree node.
+ * function TreeNode(val, left, right) {
+ *     this.val = (val===undefined ? 0 : val)
+ *     this.left = (left===undefined ? null : left)
+ *     this.right = (right===undefined ? null : right)
+ * }
+ */
+/**
+ * BFS | LevelOrderTraversal
+ * Time O(n) | Space O(n)
+ * https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/
+ * @param {TreeNode} root
+ * @return {number[][]}
+ */
+var zigzagLevelOrder = function(root) {
+    
+    const q = new Queue();
+
+    if(root) q.enqueue(root);
+
+    let isLeft = true;
```

Comment:
Can use better names

```js

/**
 * BFS
 * Time O(N) | Space O(N)
 * https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = (root) => {
    const isEdgeBase = (root === null);
    if (isEdgeBase) return [];
    
    return search(root);/* Time O(N) | Space O(N) */
};

var search = (root, isZigZag = true, order = []) => {
    const queue = new Queue([ root ]);
    
    while (!queue.isEmpty()) {        /* Time O(N) */
        const levels = [];

        bfs(queue, isZigZag, levels); /* Time O(WIDTH) | Space O(WIDTH) */
        order.push(levels);           /*                 Space O(N) */
        isZigZag = !isZigZag;
    }
    
    return order;
}

const bfs = (queue, isZigZag, levels) => {
    for (let level = queue.size(); (0 < level); level--) {/* Time O(WIDTH) */
        const { left, val, right } = queue.dequeue();
    
        if (left) queue.enqueue(left);  /* Space O(WIDTH) */
        if (right) queue.enqueue(right);/* Space O(WIDTH) */
        
        levels.push(val);               /* Space O(N) */
    }

    if (!isZigZag) levels.reverse();
}

/**
 * DFS - Pre-Order
 * Time O(N) | Space O(N)
 * https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = (root, level = 0, order = []) => {
    const isBaseCase = (root === null);
    if (isBaseCase) return [];

    return search(root, level, order);
}

var search = (node, level, order) => {
    const isEven = ((level % 2) === 0);
    const action = isEven ? 'push' : 'unshift';
    const newLevel = (order[level] ?? []);

    newLevel[action](node.val);
    order[level] = newLevel;

    return dfsPreOrder(node, level, order);
}

const dfsPreOrder = (node, level, order) => {
    const { left, right } = node;

    zigzagLevelOrder(left, (level + 1), order);  /* Time O(N) | Space O(N) */
    zigzagLevelOrder(right, (level + 1), order);/* Time O(N) | Space O(N) */

    return order;
}
```

Reply:
Please take a look at the updated code.
submission link for the proposed code: https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/submissions/1388722933/

```javascript
/**
 * BFS
 * Time O(N) | Space O(N)
 * https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = (root) => {
    const isEdgeBase = (root === null);
    if (isEdgeBase) return [];
    
    return search(root);/* Time O(N) | Space O(N) */
};

var search = (root, isZigZag = true, order = []) => {
    const queue = new Queue([ root ]);
    
    while (!queue.isEmpty()) {        /* Time O(N) */
        const levels = [];

        bfs(queue, isZigZag, levels); /* Time O(WIDTH) | Space O(WIDTH) */
        order.push(levels);           /*                 Space O(N) */
        isZigZag = !isZigZag;
    }
    
    return order;
}

const bfs = (queue, isZigZag, levels) => {
    for (let level = queue.size(); (0 < level); level--) {/* Time O(WIDTH) */
        const { left, val, right } = queue.dequeue();
    
        if (left) queue.enqueue(left);  /* Space O(WIDTH) */
        if (right) queue.enqueue(right);/* Space O(WIDTH) */
        
        levels.push(val);               /* Space O(N) */
    }

    if (!isZigZag) levels.reverse();
}


```

Reply:
@aakhtar3 (tagging to draw attention to the proposed code)

Reply:
@aadil42 Your proposed code looks good. Spacing is also good for this one. Make sure to change your current submission file to reflect your updated code

Reply:
@Ykhan799, I updated the code.